File: load-tests/src/data/manifest.ts
declare function open(path: string): string;

export interface ManifestUser {
  id: string;
  username: string;
  role: 'admin' | 'teacher' | 'student';
  full_name: string;
}

export interface ManifestClass {
  id: string;
  title: string;
  grade_level: string | null;
  is_deleted: boolean;
}

export interface ManifestAssessment {
  id: string;
  class_id: string;
  title: string;
  total_points: number;
  is_published: boolean;
  grading_period_number: number;
  question_ids: string[];
}

export interface ManifestAssignment {
  id: string;
  class_id: string;
  title: string;
  total_points: number;
  is_published: boolean;
  grading_period_number: number;
}

export interface SeedManifest {
  users: ManifestUser[];
  classes: ManifestClass[];
  assessments: ManifestAssessment[];
  assignments: ManifestAssignment[];
}

// TODO: Populate seed-manifest.json by running:
//   ./server seed-manual --export-manifest ../load-tests/seed-manifest.json
// Until then this will throw at k6 init time.
const raw = open('./seed-manifest.json');
export const manifest: SeedManifest = JSON.parse(raw);

export const publishedAssessments = manifest.assessments.filter(
  (a) => a.is_published,
);

export const publishedAssignments = manifest.assignments.filter(
  (a) => a.is_published,
);

export const activeClasses = manifest.classes.filter(
  (c) => !c.is_deleted,
);

export const students = manifest.users.filter(
  (u) => u.role === 'student',
);

export const teachers = manifest.users.filter(
  (u) => u.role === 'teacher',
);

export function randomItem<T>(arr: T[]): T {
  return arr[Math.floor(Math.random() * arr.length)];
}

File: server/src/seed/manifest.rs
use serde::Serialize;
use uuid::Uuid;

use crate::seed::specs::{AssessmentSpec, AssignmentSpec, ClassSpec, UserSpec};

#[derive(Debug, Serialize)]
pub struct ManifestUser {
    pub id: Uuid,
    pub username: String,
    pub role: String,
    pub full_name: String,
}

#[derive(Debug, Serialize)]
pub struct ManifestClass {
    pub id: Uuid,
    pub title: String,
    pub grade_level: Option<String>,
    pub is_deleted: bool,
}

#[derive(Debug, Serialize)]
pub struct ManifestAssessment {
    pub id: Uuid,
    pub class_id: Uuid,
    pub title: String,
    pub total_points: i32,
    pub is_published: bool,
    pub grading_period_number: i32,
    pub question_ids: Vec<Uuid>,
}

#[derive(Debug, Serialize)]
pub struct ManifestAssignment {
    pub id: Uuid,
    pub class_id: Uuid,
    pub title: String,
    pub total_points: i32,
    pub is_published: bool,
    pub grading_period_number: i32,
}

#[derive(Debug, Serialize)]
pub struct SeedManifest {
    pub users: Vec<ManifestUser>,
    pub classes: Vec<ManifestClass>,
    pub assessments: Vec<ManifestAssessment>,
    pub assignments: Vec<ManifestAssignment>,
}

pub fn build_manifest(
    users: &[UserSpec],
    classes: &[ClassSpec],
    assessments: &[AssessmentSpec],
    assignments: &[AssignmentSpec],
) -> SeedManifest {
    SeedManifest {
        users: users
            .iter()
            .filter(|u| u.deleted_at.is_none())
            .map(|u| ManifestUser {
                id: u.id,
                username: u.username.clone(),
                role: u.role.clone(),
                full_name: u.full_name.clone(),
            })
            .collect(),

        classes: classes
            .iter()
            .map(|c| ManifestClass {
                id: c.id,
                title: c.title.clone(),
                grade_level: c.grade_level.clone(),
                is_deleted: c.deleted_at.is_some(),
            })
            .collect(),

        assessments: assessments
            .iter()
            .filter(|a| a.deleted_at.is_none())
            .map(|a| ManifestAssessment {
                id: a.id,
                class_id: a.class_id,
                title: a.title.clone(),
                total_points: a.total_points,
                is_published: a.is_published,
                grading_period_number: a.grading_period_number,
                question_ids: a.questions.iter().map(|q| q.id).collect(),
            })
            .collect(),

        assignments: assignments
            .iter()
            .filter(|a| a.deleted_at.is_none())
            .map(|a| ManifestAssignment {
                id: a.id,
                class_id: a.class_id,
                title: a.title.clone(),
                total_points: a.total_points,
                is_published: a.is_published,
                grading_period_number: a.grading_period_number,
            })
            .collect(),
    }
}

pub fn export_manifest(manifest: &SeedManifest, path: &str) -> Result<(), Box<dyn std::error::Error>> {
    let json = serde_json::to_string_pretty(manifest)?;
    std::fs::write(path, json)?;
    println!("Seed manifest exported to: {}", path);
    Ok(())
}
